Display message when a ratelimit is hit

lordwelch opened the pull request.

lordwelch committed:
Display message when a ratelimit is hit

mizaki commented on comictaggerlib/taggerwindow.py:
```diff
@@ -1088,31 +1092,74 @@ def query_online(self, autoselect: bool = False, literal: bool = False) -> None:
+
+    def on_ratelimit(self, full_time: float, sleep_time: float) -> None:
+        toast = Toast(self)
+        toast.setDuration(10000)
```
Use `full_time` for the duration?

lordwelch replied:
It'll get triggered again on sleep_time

lordwelch committed:
Add tests

lordwelch committed:
Display toast notification longer

lordwelch committed:
Skip GUI tests on Windows and Linux

mizaki commented on comictaggerlib/issueselectionwindow.py:
```diff
@@ -201,31 +140,19 @@ def update_row(self, row: int, issue: GenericMetadata) -> None:
             try:
-                issue = self.talker.fetch_comic_data(issue_id=self.issue_id)
+                issue = self.talker.fetch_comic_data(
+                    issue_id=self.issue_id, on_rate_limit=RLCallBack(self.ratelimit, 10)
```
`self.on_ratelimit` typo?

mizaki commented on comictaggerlib/seriesselectionwindow.py:
```diff
@@ -200,6 +317,28 @@ def filter(self, text: str) -> None:
+        ):
+            QtWidgets.QApplication.setOverrideCursor(QtGui.QCursor(QtCore.Qt.CursorShape.WaitCursor))
+            try:
+                series = self.talker.fetch_series(self.series_id, on_rate_limit=RLCallBack(self.ratelimit, 10))
```
`self.on_ratelimit` typo?

mizaki commented:
Did you want to do something like this with `taggerwindow`->`on_ratelimit` or do the darkmode check once at start?
```python
palette: QtGui.QPalette = self.palette()
window_color = palette.color(QtGui.QPalette.ColorRole.Window)
is_darkmode = True if window_color.lightness() < 100 else False
if is_darkmode:
    toast.applyPreset(ToastPreset.WARNING_DARK)
else:
    toast.applyPreset(ToastPreset.WARNING)
```

lordwelch committed:
Fix dark mode

lordwelch committed:
Fix typo

mizaki commented on comictaggerlib/autotagmatchwindow.py:
```diff
@@ -89,6 +86,9 @@ def __init__(
 
         self.update_data()
 
+    def fetch(self, match: IssueResult) -> GenericMetadata:
+        return self.current_talker().fetch_comic_data(issue_id=match.issue_id)
+
```
It looks like `self.current_talker()` needs to be changed to `self.talker` and `self.talker = talker` in the `init`

mizaki replied:
Also, it's not visible here but `save_match` L:230 `md, error = self.parent().read_selected_tags(self._tags, ca)` needs to be changed to `md, _, error = self.parent().read_selected_tags(self._tags, ca)`

I would guess the same is true for `renamewindow.py` at L:85

mizaki commented:
Using the autotag window, if you cancel when the limit is waiting, the cancel (showing other windows) doesn't trigger until the wait time is over, even if the GUI wait window is closed.

lordwelch committed:
Improve canceling during a ratelimit

lordwelch commented:
I don't know of a way to fix that in python. In Go that's easy [comic-hasher/cv/cv.go][ch] but I don't know how to notify a sleeping thread in python.

[ch]: https://gitea.narnian.us/lordwelch/comic-hasher/src/commit/260a13688a69746ba5ed6626dbda5d2390ad4d31/cv/cv.go#L492-L496

mizaki commented:
> I don't know of a way to fix that in python. In Go that's easy [comic-hasher/cv/cv.go](https://gitea.narnian.us/lordwelch/comic-hasher/src/commit/260a13688a69746ba5ed6626dbda5d2390ad4d31/cv/cv.go#L492-L496) but I don't know how to notify a sleeping thread in python.

It shouldn't be too much of an issue as the wait should only be a few seconds most of the time? I upped it to 60 seconds wait time for testing.

lordwelch commented:
It gets kind of long when you hit the 200/hour limit

mizaki commented:
This might be another "not easy" but is there anyway to update the limiter if there is a time sent back by the server, i.e (`x-ratelimit-retry-after`?

lordwelch commented:
No, you'd have to basically do a whole rewrite
But that's easy to do outside of this rate-limiting it tells you how long you need to wait

mizaki commented:
Triggering the limiter with the issue window causes:
```
2025-01-09T00:54:23 | comictagger | CRITICAL | Uncaught exception: AttributeError: 'IssueSelectionWindow' object has no attribute 'iddialog'
Traceback (most recent call last):
  File "/home/workspace/comictagger/comictaggerlib/seriesselectionwindow.py", line 270, in cell_double_clicked
    self.show_issues()
  File "/home/workspace/comictagger/comictaggerlib/seriesselectionwindow.py", line 452, in show_issues
    selector = IssueSelectionWindow(self, self.config, self.talker, self.series_id, self.issue_number)
  File "/home/workspace/comictagger/comictaggerlib/issueselectionwindow.py", line 63, in __init__
    self.perform_query()
  File "/home/workspace/comictagger/comictaggerlib/issueselectionwindow.py", line 84, in perform_query
    for x in self.talker.fetch_issues_in_series(
  File "/home/workspace/comictagger/comictalker/talkers/comicvine.py", line 420, in fetch_issues_in_series
    for x in self._fetch_issues_in_series(
  File "/home/workspace/comictagger/comictalker/talkers/comicvine.py", line 848, in _fetch_issues_in_series
    cv_response: CVResult[list[CVIssue]] = self._get_cv_content(
  File "/home/workspace/comictagger/comictalker/talkers/comicvine.py", line 719, in _get_cv_content
    with self.limiter.ratelimit(
  File "/home/workspace/comictagger/comictalker/vendor/pyrate_limiter/limit_context_decorator.py", line 70, in __enter__
    self.delayed_acquire()
  File "/home/workspace/comictagger/comictalker/vendor/pyrate_limiter/limit_context_decorator.py", line 95, in delayed_acquire
    self.on_rate_limit.callback(full_delay_time, delay_time)
  File "/home/workspace/comictagger/comictaggerlib/seriesselectionwindow.py", line 273, in on_ratelimit
    self.log_output(
  File "/home/workspace/comictagger/comictaggerlib/seriesselectionwindow.py", line 392, in log_output
    if self.iddialog is not None:
AttributeError: 'IssueSelectionWindow' object has no attribute 'iddialog'
```
The `class IssueSelectionWindow(SeriesSelectionWindow):` inits with `super().__init__(parent, config, talker)` and so doesn't create the `self.iddialog` because the `series_name` (L:206) is empty.

lordwelch committed:
Use signals a little better and avoid QDialog.exec

lordwelch commented:
That should be better!

lordwelch committed:
Fix toasts and modal dialogs

Toasts calculated the duration bar in python this is now a QPropertyAnimation

Series/Issue Selection windows now use signals/slots to communicate

mizaki commented:
Getting the issue window stuck (using 3 requests a minute) and no toast. Then after it's finished, two toasts.
Initial error: `QObject::startTimer: Timers cannot be started from another thread`

Video:

https://github.com/user-attachments/assets/8629b522-06b6-4940-97f2-222d1a3912f4

lordwelch commented:
Not sure what timer is causing that... It works much better on macos, I'll have to test some stuff on linux

mizaki commented:
With the Series Window when the limiter triggers it doesn't give the `QObject::startTimer: Timers cannot be started from another thread` error. It locks the GUI and then the toast appears after the timer is over the same as the issue window.

Because it's in a thread, it's going to be a pain to debug?

lordwelch commented:
I probably have to do some bullshit with a new qobject and moveToThread

lordwelch committed:
Set Toast to use the Popup window hint

lordwelch committed:
Comment qt functions executed in a separate thread

lordwelch commented:
@mizaki try this and see if it still hangs

mizaki commented:
That seems to have sorted the fetch issue list.

Triggering the limit by clicking each issue after still has the problem but I guess you'd expect that going by the fix.

lordwelch committed:
Better event driven dialogs. Update pyqttoast for pyqt6

lordwelch committed:
Cleanup more dialogs

lordwelch committed:
Start removal of QMessageBox static methods

mizaki commented:
> Using the autotag window, if you cancel when the limit is waiting, the cancel (showing other windows) doesn't trigger until the wait time is over, even if the GUI wait window is closed.

As a (not nice) workaround, make the autotag window non-closable when the toast is active and make the toast non-closable? If it's not modal, user should still be able to click cancel in the autotag window?

lordwelch commented:
There's actually options for canceling timers/sleep https://docs.python.org/3/library/threading.html#event-objects

I need to finish cleaning up the threads/signals/slots though

lordwelch committed:
Add links to issues and series

lordwelch committed:
Fix some modal issues with dialogs

lordwelch committed:
Fix credit window

lordwelch committed:
Fix ratelimit toast position

mizaki commented:
Trying the latest changes and getting an error when using the test in GUI talker settings:
```
Original exception was:
Traceback (most recent call last):
  File "/home/bob/workspace/comictagger/comictaggerlib/ui/talkeruigenerator.py", line 80, in call_check_api
    return qtutils.information(TalkerTab.tab, "API Test Success", check_text)
  File "/home/bob/workspace/comictagger/comictaggerlib/ui/qtutils.py", line 255, in information
    qmsg = QtWidgets.QMessageBox(parent)
TypeError: arguments did not match any overloaded call:
  QMessageBox(parent: Optional[QWidget] = None): argument 1 has unexpected type '_collections._tuplegetter'
  QMessageBox(icon: QMessageBox.Icon, title: Optional[str], text: Optional[str], buttons: QMessageBox.StandardButton = QMessageBox.NoButton, parent: Optional[QWidget] = None, flags: Qt.WindowType = Qt.Dialog|Qt.MSWindowsFixedSizeDialogHint): argument 1 has unexpected type '_collections._tuplegetter'
```

mizaki commented:
Is it possible to limit the toast to the CT window? Otherwise, if you're in another window it will pop up over the top.

lordwelch committed:
Fix passing in parent

lordwelch commented:
88f11cd doesn't seem to work :-(

lordwelch committed:
Use system notifications if available

lordwelch committed:
Only enable system tray when sending a notification because QT sucks

mizaki commented:
It seems as if autotag goes through the motions but never actually writes the tags.

Also, when the manual pick window for unsure results came up and I picked and accepted:
```
Traceback (most recent call last):
  File "/home/bob/workspace/comictagger/comictaggerlib/autotagmatchwindow.py", line 193, in accept
    self.save_match()
  File "/home/bob/workspace/comictagger/comictaggerlib/autotagmatchwindow.py", line 228, in save_match
    md, error = self.parent().read_selected_tags(self._tags, ca)
ValueError: too many values to unpack (expected 2)
Error in sys.excepthook:
Traceback (most recent call last):
  File "/home/bob/workspace/comictagger/comictaggerlib/gui.py", line 68, in exception_hook
    self._exception_caught.emit(f"Oops. An unexpected error occurred:\n{log_msg}", trace_back)
TypeError: UncaughtHook._exception_caught[object] signal has 1 argument(s) but 2 provided

Original exception was:
Traceback (most recent call last):
  File "/home/bob/workspace/comictagger/comictaggerlib/autotagmatchwindow.py", line 193, in accept
    self.save_match()
  File "/home/bob/workspace/comictagger/comictaggerlib/autotagmatchwindow.py", line 228, in save_match
    md, error = self.parent().read_selected_tags(self._tags, ca)
ValueError: too many values to unpack (expected 2)
```

lordwelch commented:
> It seems as if autotag goes through the motions but never actually writes the tags.

That's just the filelist not updating

lordwelch committed:
Fix #791

lordwelch committed:
More event handling

mizaki commented:
It seems the covers aren't showing in the autotag window, neither local more remote.

mizaki commented on comictaggerlib/autotagprogresswindow.py:
```diff
@@ -17,16 +17,207 @@
+            assert res.md
+
+            def write_Tags(ca: ComicArchive, md: GenericMetadata) -> bool:
+                for tag_id in self.config.Runtime_Options__tags_write:
```
I changed this to `internal__write_tags` and autotag now writes tags.
taggerwindow:L258 takes care of copying `Runtime_Options__tags_write` into `internal__write_tags` for ref

mizaki commented:
> It seems the covers aren't showing in the autotag window, neither local more remote.

I managed to get this error which seems likely cause:
```
2025-09-29T16:28:09 | comictagger                    | CRITICAL | Uncaught exception: NameError: name 'QtNetwork' is not defined
Traceback (most recent call last):
  File "/home/bob/workspace/comictagger/comictaggerlib/imagefetcher.py", line 123, in finish_request
    length = reply.header(QtNetwork.QNetworkRequest.KnownHeaders.ContentLengthHeader)
NameError: name 'QtNetwork' is not defined
Error in sys.excepthook:
Traceback (most recent call last):
  File "/home/bob/workspace/comictagger/comictaggerlib/gui.py", line 68, in exception_hook
    self._exception_caught.emit(f"Oops. An unexpected error occurred:\n{log_msg}", trace_back)
TypeError: UncaughtHook._exception_caught[object] signal has 1 argument(s) but 2 provided
```

mizaki commented:
The being able to set the wait time didn't make it in? People are going to see `Rate limit reached: 94s until next request. Waiting 60s for ratelimit` and wonder why not wait the 94s

mizaki commented:
If you cancel the autotag there is no summary window, is that to get around it popping up only after the limiter has run down?

lordwelch committed:
Fixes

Fix qt imports
Better cancelation

lordwelch commented:
> The being able to set the wait time didn't make it in?

Nope, that will have to happen later

> People are going to see `Rate limit reached: 94s until next request. Waiting 60s for ratelimit` and wonder why not wait the 94s

I want it obvious that it's waiting for a ratelimit, when it's a manual search or auto-identify it's every 10s
>If you cancel the autotag there is no summary window, is that to get around it popping up only after the limiter has run down?

Fixed

> It seems the covers aren't showing in the autotag window, neither local more remote.

Fixed
